core/utils.c

```diff
@@ -960,9 +965,16 @@ mutex_delete(mutex_t *lock)
          * as they're deleted) and then walk it from dynamo_exit_post_detach().
          */
         lock->count_times_acquired = 0;
+# ifdef CLIENT_INTERFACE
+        skip_lock_request_assert = lock->app_lock;
+# endif /* CLIENT_INTERFACE */
     }
-#endif
-    ASSERT(lock->lock_requests == LOCK_FREE_STATE);
+#else
+# ifdef CLIENT_INTERFACE
+    skip_lock_request_assert = doing_detach;
```

Comment:
Hmm, not sure about this: cleanup of regular locks during detach should still involve a clean lock state, right?

Reply:
How can we tell whether this is a "regular lock" here rather than an app lock? This block is if DEADLOCK_AVOIDANCE is not defined, so the mutex doesn't have an app_lock field. My attempt to narrow the scope was putting in the CLIENT_INTERFACE check, but I agree the current logic skips more cases than it should.

Reply:
Today DEADLOCK_AVOIDANCE is only defined for DEBUG which corresponds to when the ASSERT will fire so we won't need this -- maybe just put a comment that app locks will hit the assert in the unsupported combo of DEADLOCK_AVOIDANCE and !DEBUG (or even put #error for that)

Reply:
I don't see how your suggestion helps here: we need to know whether to skip the ASSERT when `DEBUG && !DEADLOCK_AVOIDANCE`. Unless `DEBUG` implies `!DEADLOCK_AVOIDANCE` too, we support a compilation mode where ASSERTs execute but the `lock->app_lock` field does not exist.

Reply:
We do not support `DEBUG && !DEADLOCK_AVOIDANCE` today so it seems better to bail there (do nothing w/ comment, or #error) rather than over-relax the assert

Reply:
Ah ok, thanks for the thoughts.